### PR TITLE
fit_locations_separately

### PR DESCRIPTION
**Pull request by elray1**
code updates and resulting model outputs for a gbq variation that fits/predicts separately by location

### CODE REVIEW

File: code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py
Comment (elray1) [inline code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:1]: for the manuscript, i think there will only be this script and one more like it added to this retrospective-experiments folder.  This is heading to a place where it should be refactored, but i'm being lazy for now, anticipating that we may land on some revisions to this experimental infrastructure/setup in a couple of weeks when we meet with collaborators.

File: code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This script generates predictions for the gbq_qr_fit_locations_separately model for all reference dates.
+# Retrospective model fits are generated using the data that would have been available in
+# real time.
+#
+# To maintain transparency about which model outputs were and were not generated in
+# real time, these model outputs are stored in flusion/retrospective-hub.
+
+# This script should be run with code/gbq as the working directory:
+# python retrospective-experiments/gbq_qr_fit_locations_separately.py
```
Comment (bsweger) [inline code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:9]: As a heads-up, I got a lot of multiprocessing-related errors when running this script on my macbook (Macbook M3). Below is an abbreviated snippet. I made a change based on a 

```
flusion/code/gbq on  fit_locations_separately [$?] via 🅒 flusion
➜
python retrospective-experiments/gbq_qr_fit_locations_separately.py
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 329, in _repopulate_pool_static
    self._repopulate_pool()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 306, in _repopulate_pool
    w.start()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/process.py", line 121, in start
    return self._repopulate_pool_static(self._ctx, self.Process,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 329, in _repopulate_pool_static
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    w.start()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/process.py", line 121, in start
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    super().__init__(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    self._launch(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    super().__init__(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 158, in get_preparation_data
    self._launch(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    _check_not_importing_main()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 138, in _check_not_importing_main
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 158, in get_preparation_data
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    _check_not_importing_main()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 138, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 120, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 129, in _main
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 120, in spawn_main
    prepare(preparation_data)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 240, in prepare
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 129, in _main
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 291, in _fixup_main_from_path
    prepare(preparation_data)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 240, in prepare
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 291, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/rsweger/code/flusion/code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py", line 30, in <module>
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 291, in _fixup_main_from_path
    with Pool(processes=2) as pool:
         ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 119, in Pool
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 291, in run_path
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/rsweger/code/flusion/code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py", line 30, in <module>
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 215, in __init__
    with Pool(processes=2) as pool:
         ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 119, in Pool
    self._repopulate_pool()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 306, in _repopulate_pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 215, in __init__
    return self._repopulate_pool_static(self._ctx, self.Process,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 329, in _repopulate_pool_static
    self._repopulate_pool()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 306, in _repopulate_pool
    w.start()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/process.py", line 121, in start
    return self._repopulate_pool_static(self._ctx, self.Process,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/pool.py", line 329, in _repopulate_pool_static
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    w.start()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/process.py", line 121, in start
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    super().__init__(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    self._launch(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    super().__init__(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 158, in get_preparation_data
    self._launch(process_obj)
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    _check_not_importing_main()
  File "/Users/rsweger/miniconda3/envs/flusion/lib/python3.11/multiprocessing/spawn.py", line 138, in _check_not_importing_main
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py
Comment (bsweger) [reply code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:1]: Makes sense--this kind of context is super helpful from a reviewer perspective, thanks for including it.

File: code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py (hunk 1 above)
Comment (elray1) [reply code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:9]: although i don't understand why it is necessary here, I confirmed that if I add line `freeze_support()` line, this script still runs on my desktop and produces the same outputs.  propose that I go ahead and add that line and we can merge this.

Comment (elray1) [reply code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:9]: (i have made that change)

File: code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        for i in range(29)]
+
+output_root = '../../retrospective-hub/model-output'
+
+commands = [f'python gbq.py --ref_date {ref_date} --output_root {output_root} --model_name gbq_qr_fit_locations_separately' \
+                for ref_date in missing_ref_dates]
+
+with Pool(processes=2) as pool:
+    freeze_support()
+    pool.map(run_command, commands)
```
Comment (bsweger) [inline code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:32]:
```suggestion
def main():
    with Pool(processes=2) as pool:
        pool.map(run_command, commands)

if __name__ == '__main__':
    freeze_support()
    main()
```

Comment (bsweger) [reply code/gbq/retrospective-experiments/gbq_qr_fit_locations_separately.py:32]: this is the incantation I used to get it working on my laptop (sorry for not being more specific earlier...just figured out how to suggest multi-line changes)

**Comment by elray1** [comment]
I committed the suggested updates and confirmed that they don't impact run outputs on my computer.

**Pull request merged**